### PR TITLE
✨  Remove `/income/details` and revamp `/income` API

### PR DESCRIPTION
**Pull request by WeiJun0827**
Response sample:
```
{
  "total_sales": 6712108800000,
  "class_incomes": [
    {
      "class_id": "likenft14tp0n69q8wwvyd6k6fv4cp4jd86r072at6pa2yesue22qtrsnxestryg89",
      "created_at": "2022-07-24T13:02:37Z",
      "sales": 4600034000000,
      "incomes": [
        {
          "address": "like1shkl5gqzxcs9yh3qjdeggaz3yg5s83754dx2dh",
          "amount": 2780800000000
        },
        {
          "address": "like1yney2cqn5qdrlc50yr5l53898ufdhxafqz9gxp",
          "amount": 1819234000000
        }
      ]
    },
    ...
  ],
  "pagination": {
    "next_key": 15,
    "count": 15
  }
}
```

### CODE REVIEW

File: db/nft.go [hunk 1]
```diff
@@ -381,135 +381,91 @@ func GetNftEvents(conn *pgxpool.Conn, q QueryEventsRequest, p PageRequest) (Quer
...
+					ON e.class_id = i.class_id
+					AND e.nft_id = i.nft_id
+					AND e.tx_hash = i.tx_hash
+				WHERE ($3 = '' OR e.class_id = $3)
+					AND ($4::text[] IS NULL OR cardinality($4::text[]) = 0 OR iscn.owner = ANY($4))
+					AND ($5 = 0 OR (e.timestamp IS NOT NULL AND e.timestamp > to_timestamp($5)))
+					AND ($6 = 0 OR (e.timestamp IS NOT NULL AND e.timestamp < to_timestamp($6)))
+					AND ($7::text[] IS NULL OR cardinality($7::text[]) = 0 OR e.action = ANY($7))
+					AND ($8 = false OR i.address = iscn.owner)
+					AND ($9 = true OR e.receiver != iscn.owner)
```
Comment (williamchong) [inline db/nft.go:418]: Self purchase would actually incur income to other stakeholders too though, so use with caution.
What is the use case for this?

File: db/nft.go [hunk 2]
```diff
@@ -381,135 +381,91 @@ func GetNftEvents(conn *pgxpool.Conn, q QueryEventsRequest, p PageRequest) (Quer
...
-	}
-	for rows.Next() {
-		var r NftIncomeDetailResponse
-		if err = rows.Scan(
-			&r.ClassId, &r.NftId, &r.TxHash, &r.Timestamp, &r.Address,
-			&r.Price, &r.Amount,
-		); err != nil {
-			logger.L.Errorw("failed to scan nft income details", "error", err, "q", q)
-			return QueryIncomeDetailsResponse{}, fmt.Errorf("query nft income details data failed: %w", err)
+		if ciPtr == nil || ci.ClassId != ciPtr.ClassId { // is new class
```
Comment (williamchong) [inline db/nft.go:455]: instead of using pointer, can't we just make a map and assign/append data accordingly?

File: db/nft.go (hunk 1 above)
Comment (WeiJun0827) [reply db/nft.go:418]: When using Kin's address as an input param, I found the first place on the income list is '來自舊金山的明信片'. I guess he probably doesn't want to count the self-purchase case?

Comment (WeiJun0827) [reply db/nft.go:418]: Or alternatively, we can change `IncludeSelfPurchase` to `ExcludeSelfPurchase`?

File: db/nft.go (hunk 2 above)
Comment (WeiJun0827) [reply db/nft.go:465]: Tried using a `rankMap := make(map[int]NftClassIncomeResponse)` to collect the row data. When converting the map to response data, it seems the iteration order of `range` operation on `map` is not guaranteed in Go. Thus [additional sorting would be required](https://stackoverflow.com/a/18342865/7978205) and make time complexity change from O(N) to O(NlogN)?

Comment (williamchong) [reply db/nft.go:466]: @nnkken please comment on how to write this best

Comment (nnkken) [reply db/nft.go:462]: better rename to `ciPrev`

Comment (nnkken) [reply db/nft.go:462]: using a map will be clearer, here the data size is small (hundreds at most I guess?) so a sorting is acceptable.
We may also have an extra array of map keys, and append to this array only if it does not exist in the map, to maintain insert order.

Comment (WeiJun0827) [reply db/nft.go]: Refactored

**Review by nnkken — APPROVED** [review]

File: utils/events.go [hunk 3]
```diff
@@ -97,12 +103,45 @@ func GetIncomeMap(events types.StringEvents) map[string]uint64 {
 					amount = coin.Amount.Uint64()
 				}
 				if address != "" && amount != 0 {
-					incomeMap[address] += amount
+					if amount > maxAmount {
+						maxAmount = amount
```
Comment (williamchong) [inline utils/events.go:107]: What is use of `maxAmount` here?

File: db/nft.go [hunk 4]
```diff
@@ -381,135 +381,144 @@ func GetNftEvents(conn *pgxpool.Conn, q QueryEventsRequest, p PageRequest) (Quer
...
+		rankMap[rank] = classIncome
+		res.TotalAmount += i.Amount
 	}
+	var classIds []string
+	for _, classIncome := range rankMap {
+		res.ClassIncomes = append(res.ClassIncomes, classIncome)
+		classIds = append(classIds, classIncome.ClassId)
+	}
+	for i := 0; i < len(res.ClassIncomes); i++ {
+		res.ClassIncomes[i] = rankMap[i+rankStart+1]
```
Comment (williamchong) [inline db/nft.go:478]:
```suggestion
		res.ClassIncomes[i] = rankMap[rankStart+i+1]
```
Personally I think this is more readable but just my preference...

**Review by williamchong — CHANGES_REQUESTED** [review]

**Review by williamchong — APPROVED** [review]

**Pull request merged**